lib/modem_trace/modem_trace.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <kernel.h>
+#include <zephyr.h>
+#include <modem/modem_trace.h>
+#include "nrf_modem_at.h"
+#include "nrfx_uarte.h"
+#include "SEGGER_RTT.h"
+
+static const nrfx_uarte_t uarte_inst = NRFX_UARTE_INSTANCE(1);
+
+static int trace_rtt_channel;
+static char rtt_buffer[CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_RTT_BUF_SIZE];
+
+static uint32_t max_trace_size_bytes;
+static uint32_t total_trace_size_rcvd;
+static bool is_transport_initialized;
+
+static void trace_stop_timer_handler(struct k_timer *timer);
+
+K_TIMER_DEFINE(trace_stop_timer, trace_stop_timer_handler, NULL);
+
+static void trace_stop_timer_handler(struct k_timer *timer)
+{
+	modem_trace_stop();
+}
+
+static bool uart_init(void)
+{
+	const nrfx_uarte_config_t config = {
+		.pseltxd = DT_PROP(DT_NODELABEL(uart1), tx_pin),
+		.pselrxd = DT_PROP(DT_NODELABEL(uart1), rx_pin),
+		.pselcts = NRF_UARTE_PSEL_DISCONNECTED,
+		.pselrts = NRF_UARTE_PSEL_DISCONNECTED,
+
+		.hal_cfg.hwfc = NRF_UARTE_HWFC_DISABLED,
+		.hal_cfg.parity = NRF_UARTE_PARITY_EXCLUDED,
+		.baudrate = NRF_UARTE_BAUDRATE_1000000,
+
+		.interrupt_priority = NRFX_UARTE_DEFAULT_CONFIG_IRQ_PRIORITY,
+		.p_context = NULL,
+	};
+
+	return (nrfx_uarte_init(&uarte_inst, &config, NULL) == NRFX_SUCCESS);
+}
+
+static bool rtt_init(void)
+{
+	trace_rtt_channel = SEGGER_RTT_AllocUpBuffer("modem_trace", rtt_buffer, sizeof(rtt_buffer),
+						     SEGGER_RTT_MODE_NO_BLOCK_SKIP);
+
+	return (trace_rtt_channel > 0);
+}
+
+int modem_trace_init(void)
+{
+	if (IS_ENABLED(CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_UART)) {
+		is_transport_initialized = uart_init();
+	}
+	if (IS_ENABLED(CONFIG_NRF_MODEM_LIB_TRACE_MEDIUM_RTT)) {
+		is_transport_initialized = rtt_init();
+	}
+
+	if (!is_transport_initialized) {
+		return -EBUSY;
+	}
+	return 0;
+}
+
+int modem_trace_start(enum modem_trace_mode trace_mode, uint16_t duration, uint32_t max_size)
+{
+	if (!is_transport_initialized) {
+		return -ENXIO;
+	}
+
+	total_trace_size_rcvd = 0;
+
+	max_trace_size_bytes = max_size;
+
+	if (nrf_modem_at_printf("AT%%XMODEMTRACE=1,%hu", trace_mode) != 0) {
+		return -EOPNOTSUPP;
+	}
+
+	if (duration != 0) {
+		k_timer_start(&trace_stop_timer, K_SECONDS(duration), K_SECONDS(0));
+	}
+
+	return 0;
+}
+
+int modem_trace_process(const uint8_t *data, uint32_t len)
+{
+	if (!is_transport_initialized) {
+		return -ENXIO;
+	}
+
+	if (max_trace_size_bytes != 0) {
```

Comment:
Added a new test case as you suggested. It failed. And so had to change the implementation a bit. Please have a look at the last commit.

>Didn't you recently discover that it starts sending trace data in a default mode when the CONFIG_MODEM_TRACE option is enabled?

Yes. This is because the modem starts sending traces as soon as it receives the shared memory configuration with space reserved for traces. The shared memory is sent to the modem during initialization by the `nrf_modem_init()` function. This is documented (as explained in my other comment)